In [1]:
import requests
import pandas as pd
import numpy as np 
import json
from pandas.io.json import json_normalize

# Dados do Governo da Republica Federativa do BRASIL

# Correspondentes bancários no país

O arquivo disponibilizado apresenta os dados mais atuais dos pontos de atendimento dos correspondentes, por instituição financeira e por município, com a identificação dos tipos de serviços prestados, conforme descrito na Resolução 3.954.

Aviso Legal: o teor das informações é de responsabilidade da respectiva instituição/empresa, de acordo com a regulamentação em vigor - ver mais: https://dadosabertos.bcb.gov.br/dataset/correspondentes


Referência bibliográfica: https://fernandamelhem.jusbrasil.com.br/artigos/315457885/correspondentes-bancarios

# Analise Exploratoria - EDA

URL utilizada referente a API consumida

Extração dos dados consumindo a API;
Trabsformação dos dados capturados em um DataSet
Demonstração do DataSet criado

In [4]:
#Consumindo dados Via API
    #cUrl = "https://olinda.bcb.gov.br/olinda/servico/Informes_Correspondentes/versao/v1/odata/Correspondentes?$top=100&$format=json&$select=CnpjContratante,NomeContratante,CnpjCorrespondente,NomeCorrespondente,Tipo,Ordem,MunicipioIBGE,Municipio,UF,ServicosCorrespondentes,Posicao"
    #cUrl = "https://olinda.bcb.gov.br/olinda/servico/Informes_Correspondentes/versao/v1/odata/Correspondentes?$top=10000&$format=json"
    #cUrl = "https://olinda.bcb.gov.br/olinda/servico/Informes_Correspondentes/versao/v1/odata/Correspondentes?$format=json"
    #dados = pd.read_json(cUrl)
#Acessando Json baixado no dia 03/11/2019 utilizando a url :
    #https://olinda.bcb.gov.br/olinda/servico/Informes_Correspondentes/versao/v1/odata/Correspondentes?$top=1000000000000&$format=json     

dados = pd.read_json(".\Correspondentes.json")    
df = pd.DataFrame(dados) 
df.head()

@odata.context  \
0  https://was-p.bcnet.bcb.gov.br/olinda/servico/...   
1  https://was-p.bcnet.bcb.gov.br/olinda/servico/...   
2  https://was-p.bcnet.bcb.gov.br/olinda/servico/...   
3  https://was-p.bcnet.bcb.gov.br/olinda/servico/...   
4  https://was-p.bcnet.bcb.gov.br/olinda/servico/...   

                                               value  
0  {'CnpjContratante': '45441789', 'NomeContratan...  
1  {'CnpjContratante': '45441789', 'NomeContratan...  
2  {'CnpjContratante': '45441789', 'NomeContratan...  
3  {'CnpjContratante': '45441789', 'NomeContratan...  
4  {'CnpjContratante': '36321990', 'NomeContratan...

 Normalizando o campo value(composto por colunas e linhas referentes aos dados dos correspondentes)

In [5]:
dfnew = (json_normalize(df.value))
dfnew.head()

CnpjContratante                                    NomeContratante  \
0        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
1        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
2        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
3        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
4        36321990  AGORACRED S/A SOCIEDADE DE CRÉDITO, FINANCIAM...   

  CnpjCorrespondente                            NomeCorrespondente    Tipo  \
0           25134887                          H MOTOS CARIOCA LTDA    Sede   
1           25134887                          H MOTOS CARIOCA LTDA  Filial   
2           26108857       PAU BRASIL BAHIA COMERCIO DE MOTOS LTDA    Sede   
3           26105548  PAU BRASIL PERNAMBUCO COMERCIO DE MOTOS LTDA    Sede   
4           02911625                          5 IRMAOS MOVEIS LTDA    Sede   

    Ordem MunicipioIBGE    Municipio  UF      ServicosCorrespondentes  \
0  I00001       3303302      NITEROI  RJ                       Inc. V   
1  I00002       3304904  SAO GONCALO  RJ                       Inc. V   
2  I00001       2918407     JUAZEIRO  BA                       Inc. V   
3  I00001       2611101    PETROLINA  PE                       Inc. V   
4  I00001       3205002        SERRA  ES  Inc. III, Inc. V, Inc. VIII   

      Posicao  
0  31/10/2019  
1  31/10/2019  
2  31/10/2019  
3  31/10/2019  
4  31/10/2019

Criando compo link com os dados da coluna @odata.context

In [6]:
dfnew['urlserv'] = df['@odata.context'] 
dfnew.head()

CnpjContratante                                    NomeContratante  \
0        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
1        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
2        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
3        45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
4        36321990  AGORACRED S/A SOCIEDADE DE CRÉDITO, FINANCIAM...   

  CnpjCorrespondente                            NomeCorrespondente    Tipo  \
0           25134887                          H MOTOS CARIOCA LTDA    Sede   
1           25134887                          H MOTOS CARIOCA LTDA  Filial   
2           26108857       PAU BRASIL BAHIA COMERCIO DE MOTOS LTDA    Sede   
3           26105548  PAU BRASIL PERNAMBUCO COMERCIO DE MOTOS LTDA    Sede   
4           02911625                          5 IRMAOS MOVEIS LTDA    Sede   

    Ordem MunicipioIBGE    Municipio  UF      ServicosCorrespondentes  \
0  I00001       3303302      NITEROI  RJ                       Inc. V   
1  I00002       3304904  SAO GONCALO  RJ                       Inc. V   
2  I00001       2918407     JUAZEIRO  BA                       Inc. V   
3  I00001       2611101    PETROLINA  PE                       Inc. V   
4  I00001       3205002        SERRA  ES  Inc. III, Inc. V, Inc. VIII   

      Posicao                                            urlserv  
0  31/10/2019  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
1  31/10/2019  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
2  31/10/2019  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3  31/10/2019  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
4  31/10/2019  https://was-p.bcnet.bcb.gov.br/olinda/servico/...

Tamanho do DataSet

In [7]:
dfnew.shape

(387847, 12)

In [8]:
dfnew.columns

Index(['CnpjContratante', 'NomeContratante', 'CnpjCorrespondente',
       'NomeCorrespondente', 'Tipo', 'Ordem', 'MunicipioIBGE', 'Municipio',
       'UF', 'ServicosCorrespondentes', 'Posicao', 'urlserv'],
      dtype='object')

In [65]:
dfnew.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 387847 entries, 0 to 387846
Data columns (total 12 columns):
CnpjContratante            387847 non-null object
NomeContratante            387847 non-null object
CnpjCorrespondente         387847 non-null object
NomeCorrespondente         387847 non-null object
Tipo                       387847 non-null object
Ordem                      387847 non-null object
MunicipioIBGE              387367 non-null float64
Municipio                  387823 non-null object
UF                         387823 non-null object
ServicosCorrespondentes    387847 non-null object
Posicao                    387847 non-null object
urlserv                    387847 non-null object
dtypes: float64(1), object(11)
memory usage: 35.5+ MB


In [64]:
dfnew["MunicipioIBGE"] = pd.to_numeric(dfnew["MunicipioIBGE"])

In [41]:
dfnew.loc[(dfnew['MunicipioIBGE'].isna())]

CnpjContratante                                    NomeContratante  \
3033          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3316          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3320          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3321          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3322          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
...                ...                                                ...   
385355        47193149      SANTANDER LEASING S.A. ARRENDAMENTO MERCANTIL   
386587        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386727        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386768        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386769        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   

       CnpjCorrespondente                       NomeCorrespondente    Tipo  \
3033             00549675                        OCT VEICULOS LTDA    Sede   
3316             09348217          SADIF COMERCIO DE VEICULOS LTDA  Filial   
3320             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
3321             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
3322             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
...                   ...                                      ...     ...   
385355           11739439  V12 MOTORS VW COMERCIO DE VEICULOS LTDA  Filial   
386587           61189288                        MARISA LOJAS S.A.  Filial   
386727           61189288                        MARISA LOJAS S.A.  Filial   
386768           61189288                        MARISA LOJAS S.A.  Filial   
386769           61189288                        MARISA LOJAS S.A.  Filial   

         Ordem MunicipioIBGE              Municipio  UF  \
3033    I00001          None  BRASILIA (TAGUATINGA)  DF   
3316    I00002          None        BRASILIA (GAMA)  DF   
3320    I00024          None       BRASILIA (GUARA)  DF   
3321    I00026          None       BRASILIA (GUARA)  DF   
3322    I00028          None       BRASILIA (GUARA)  DF   
...        ...           ...                    ...  ..   
385355  I00003          None       BRASILIA (GUARA)  DF   
386587  I00087          None  BRASILIA (TAGUATINGA)  DF   
386727  I00237          None  BRASILIA (TAGUATINGA)  DF   
386768  I00281          None  BRASILIA (TAGUATINGA)  DF   
386769  I00282          None   BRASILIA (CEILANDIA)  DF   

       ServicosCorrespondentes     Posicao  \
3033                    Inc. V  31/10/2019   
3316                    Inc. V  31/10/2019   
3320                    Inc. V  31/10/2019   
3321                    Inc. V  31/10/2019   
3322                    Inc. V  31/10/2019   
...                        ...         ...   
385355                  Inc. V  31/10/2019   
386587                  Inc. V  31/10/2019   
386727                  Inc. V  31/10/2019   
386768                  Inc. V  31/10/2019   
386769                  Inc. V  31/10/2019   

                                                  urlserv  
3033    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3316    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3320    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3321    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3322    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
...                                                   ...  
385355  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386587  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386727  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386768  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386769  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  

[480 rows x 12 columns]

In [42]:
 dfnew.loc[(dfnew['Municipio'].isna())]

CnpjContratante                                    NomeContratante  \
22958         07707650  AYMORÉ CRÉDITO, FINANCIAMENTO E INVESTIMENTO...   
29851         61186680                                     BANCO BMG S.A.   
49184         07207996                 BANCO BRADESCO FINANCIAMENTOS S.A.   
126674        00000000                               BANCO DO BRASIL S.A.   
130813        00000000                               BANCO DO BRASIL S.A.   
149069        17192451                                BANCO ITAUCARD S.A.   
156328        17192451                                BANCO ITAUCARD S.A.   
157942        17192451                                BANCO ITAUCARD S.A.   
160923        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
160965        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
161000        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
161001        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
162791        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
162862        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
163419        33885724                        BANCO ITAÚ CONSIGNADO S.A.   
253877        01149953  BV FINANCEIRA S.A. - CREDITO, FINANCIAMENTO E ...   
312287        06881898  FINANCEIRA ITAÚ CBD S.A. CRÉDITO, FINANCIAME...   
314779        60701190                                ITAÚ UNIBANCO S.A.   
314789        60701190                                ITAÚ UNIBANCO S.A.   
317828        60701190                                ITAÚ UNIBANCO S.A.   
321524        60701190                                ITAÚ UNIBANCO S.A.   
326372        60701190                                ITAÚ UNIBANCO S.A.   
328379        60701190                                ITAÚ UNIBANCO S.A.   
339641        02682287                    PAN ARRENDAMENTO MERCANTIL S.A.   

       CnpjCorrespondente                                 NomeCorrespondente  \
22958            26033888                    RRV COMERCIO DE VEICULOS EIRELI   
29851            24546995                    GOLD MONEY MULTISSERVICOS LTDA.   
49184            26033888                    RRV COMERCIO DE VEICULOS EIRELI   
126674           28057857                                  J.A DROGARIA LTDA   
130813           32696651        MARIA CECILIA CLAUDINO DA SILVA 13677205409   
149069           23196759                       AZZURRA FRANCE VEICULOS LTDA   
156328           17352305                          NOVACAR RIO VEICULOS LTDA   
157942           26033888                    RRV COMERCIO DE VEICULOS EIRELI   
160923           12467416      CBFACIL CORRETORA DE SEGUROS E NEGOCIOS LTDA.   
160965           12467416      CBFACIL CORRETORA DE SEGUROS E NEGOCIOS LTDA.   
161000           12467416      CBFACIL CORRETORA DE SEGUROS E NEGOCIOS LTDA.   
161001           12467416      CBFACIL CORRETORA DE SEGUROS E NEGOCIOS LTDA.   
162791           10554049  NATUREZA CRED PRESTACAO DE SERVICOS DE INFORMA...   
162862           07441080                           NUNESCRED CENTRAL EIRELI   
163419           02983474  SEDAE - SERVICO DE DESENVOLVIMENTO E ADMINISTR...   
253877           32999293                DENIS RODRIGO DA CONCEICAO DE MATOS   
312287           06057223                           SENDAS DISTRIBUIDORA S/A   
314779           19998113                                AZUCAR SHOES EIRELI   
314789           23196759                       AZZURRA FRANCE VEICULOS LTDA   
317828           17256512                                DROGARIA ARAUJO S A   
321524           08662164                     JAILSON MACEDO SILVA ARMARINHO   
326372           17352305                          NOVACAR RIO VEICULOS LTDA   
328379           26033888                    RRV COMERCIO DE VEICULOS EIRELI   
339641           32999293                DENIS RODRIGO DA CONCEICAO DE MATOS   

          Tipo   Ordem MunicipioIBGE Municipio    UF  \
22958   Filial  I00003  

In [44]:
#df1 ira receber apenas linhas sem valores nulos
#df2 ira receber apenas linhas com valores nulos
df1 = dfnew.dropna()
df2 = dfnew[pd.isnull(dfnew).any(axis=1)]

In [45]:
df1

CnpjContratante                                    NomeContratante  \
0             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
1             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
2             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
3             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
4             36321990  AGORACRED S/A SOCIEDADE DE CRÉDITO, FINANCIAM...   
...                ...                                                ...   
387842        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387843        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387844        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387845        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387846        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   

       CnpjCorrespondente                            NomeCorrespondente  \
0                25134887                          H MOTOS CARIOCA LTDA   
1                25134887                          H MOTOS CARIOCA LTDA   
2                26108857       PAU BRASIL BAHIA COMERCIO DE MOTOS LTDA   
3                26105548  PAU BRASIL PERNAMBUCO COMERCIO DE MOTOS LTDA   
4                02911625                          5 IRMAOS MOVEIS LTDA   
...                   ...                                           ...   
387842           02817838                 UTILIDADES DOMESTICAS UD LTDA   
387843           02817838                 UTILIDADES DOMESTICAS UD LTDA   
387844           02817838                 UTILIDADES DOMESTICAS UD LTDA   
387845           02817838                 UTILIDADES DOMESTICAS UD LTDA   
387846           02817838                 UTILIDADES DOMESTICAS UD LTDA   

          Tipo   Ordem MunicipioIBGE        Municipio  UF  \
0         Sede  I00001       3303302          NITEROI  RJ   
1       Filial  I00002       3304904      SAO GONCALO  RJ   
2         Sede  I00001       2918407         JUAZEIRO  BA   
3         Sede  I00001       2611101        PETROLINA  PE   
4         Sede  I00001       3205002            SERRA  ES   
...        ...     ...           ...              ...  ..   
387842  Filial  I00019       3123205  DORES DO INDAIA  MG   
387843  Filial  I00021       3114204  CARMO DO CAJURU  MG   
387844  Filial  I00024       3138807              LUZ  MG   
387845  Filial  I00018       3129509             IBIA  MG   
387846  Filial  I00027       3137205   LAGOA DA PRATA  MG   

            ServicosCorrespondentes     Posicao  \
0                            Inc. V  31/10/2019   
1                            Inc. V  31/10/2019   
2                            Inc. V  31/10/2019   
3                            Inc. V  31/10/2019   
4       Inc. III, Inc. V, Inc. VIII  31/10/2019   
...                             ...         ...   
387842                       Inc. V  31/10/2019   
387843                       Inc. V  31/10/2019   
387844                       Inc. V  31/10/2019   
387845                       Inc. V  31/10/2019   
387846                       Inc. V  31/10/2019   

                                                  urlserv  
0       https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
1       https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
2       https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3       https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
4       https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
...                                                   ...  
387842  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
387843  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
387844  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
387845  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
387846  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  

[387367 rows x 12 columns]

In [46]:
df2

CnpjContratante                                    NomeContratante  \
3033          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3316          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3320          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3321          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3322          46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
...                ...                                                ...   
385355        47193149      SANTANDER LEASING S.A. ARRENDAMENTO MERCANTIL   
386587        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386727        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386768        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   
386769        07747410  SAX S.A. - CRÉDITO, FINANCIAMENTO E INVESTIMENTO   

       CnpjCorrespondente                       NomeCorrespondente    Tipo  \
3033             00549675                        OCT VEICULOS LTDA    Sede   
3316             09348217          SADIF COMERCIO DE VEICULOS LTDA  Filial   
3320             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
3321             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
3322             00752386        SAFARI COMERCIO DE VEICULOS LTDA.  Filial   
...                   ...                                      ...     ...   
385355           11739439  V12 MOTORS VW COMERCIO DE VEICULOS LTDA  Filial   
386587           61189288                        MARISA LOJAS S.A.  Filial   
386727           61189288                        MARISA LOJAS S.A.  Filial   
386768           61189288                        MARISA LOJAS S.A.  Filial   
386769           61189288                        MARISA LOJAS S.A.  Filial   

         Ordem MunicipioIBGE              Municipio  UF  \
3033    I00001          None  BRASILIA (TAGUATINGA)  DF   
3316    I00002          None        BRASILIA (GAMA)  DF   
3320    I00024          None       BRASILIA (GUARA)  DF   
3321    I00026          None       BRASILIA (GUARA)  DF   
3322    I00028          None       BRASILIA (GUARA)  DF   
...        ...           ...                    ...  ..   
385355  I00003          None       BRASILIA (GUARA)  DF   
386587  I00087          None  BRASILIA (TAGUATINGA)  DF   
386727  I00237          None  BRASILIA (TAGUATINGA)  DF   
386768  I00281          None  BRASILIA (TAGUATINGA)  DF   
386769  I00282          None   BRASILIA (CEILANDIA)  DF   

       ServicosCorrespondentes     Posicao  \
3033                    Inc. V  31/10/2019   
3316                    Inc. V  31/10/2019   
3320                    Inc. V  31/10/2019   
3321                    Inc. V  31/10/2019   
3322                    Inc. V  31/10/2019   
...                        ...         ...   
385355                  Inc. V  31/10/2019   
386587                  Inc. V  31/10/2019   
386727                  Inc. V  31/10/2019   
386768                  Inc. V  31/10/2019   
386769                  Inc. V  31/10/2019   

                                                  urlserv  
3033    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3316    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3320    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3321    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
3322    https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
...                                                   ...  
385355  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386587  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386727  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386768  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  
386769  https://was-p.bcnet.bcb.gov.br/olinda/servico/...  

[480 rows x 12 columns]

Por ter verificado que existem valores nulos no universo de 387847 registros, apenas nas colunas:

MunicipioIBGE              387367

Municipio                  387823

UF                         387823

Observando que o campo MunicipioIBGE por ter o numero maior de valores nulos pode ser populado
levando em consideração a descrição do Municipio

In [56]:
cUrlMunicipios = "https://servicodados.ibge.gov.br/api/v1/localidades/municipios?$format=json"
#"https://servicodados.ibge.gov.br/api/v1/localidades/estados/{UF}/municipios"
municipios = pd.read_json(cUrlMunicipios)
dfCities = pd.DataFrame(municipios) 
dfCities.head()

id                   nome  \
0  1100015  Alta Floresta D'Oeste   
1  1100023              Ariquemes   
2  1100031                 Cabixi   
3  1100049                 Cacoal   
4  1100056             Cerejeiras   

                                        microrregiao  
0  {'id': 11006, 'nome': 'Cacoal', 'mesorregiao':...  
1  {'id': 11003, 'nome': 'Ariquemes', 'mesorregia...  
2  {'id': 11008, 'nome': 'Colorado do Oeste', 'me...  
3  {'id': 11006, 'nome': 'Cacoal', 'mesorregiao':...  
4  {'id': 11008, 'nome': 'Colorado do Oeste', 'me...

In [57]:
dfCities.rename(columns={'id': 'MunicipioIBGE', 'nome': 'Municipio'}, inplace=True)
dfCities = dfCities.drop('microrregiao', axis=1)
dfCities.head()

MunicipioIBGE              Municipio
0        1100015  Alta Floresta D'Oeste
1        1100023              Ariquemes
2        1100031                 Cabixi
3        1100049                 Cacoal
4        1100056             Cerejeiras

In [63]:
dfCities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5570 entries, 0 to 5569
Data columns (total 2 columns):
MunicipioIBGE    5570 non-null int64
Municipio        5570 non-null object
dtypes: int64(1), object(1)
memory usage: 87.2+ KB


In [66]:
dfCities["MunicipioIBGE"] = pd.to_numeric(dfCities["MunicipioIBGE"])

In [67]:
dfnew.merge(dfCities, left_on='MunicipioIBGE', right_on='MunicipioIBGE')

CnpjContratante                                    NomeContratante  \
0             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
1             46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
2             46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
3             46570800                   ALFA ARRENDAMENTO MERCANTIL S.A.   
4             07707650  AYMORÉ CRÉDITO, FINANCIAMENTO E INVESTIMENTO...   
...                ...                                                ...   
387362        92702067          BANCO DO ESTADO DO RIO GRANDE DO SUL S.A.   
387363        00360305                            CAIXA ECONOMICA FEDERAL   
387364        92555150  COOPERATIVA DE CRÉDITO, POUPANÇA E INVESTIME...   
387365        92702067          BANCO DO ESTADO DO RIO GRANDE DO SUL S.A.   
387366        00360305                            CAIXA ECONOMICA FEDERAL   

       CnpjCorrespondente                              NomeCorrespondente  \
0                25134887                            H MOTOS CARIOCA LTDA   
1                06888977             EPC DISTRIBUIDORA DE VEICULOS LTDA.   
2                02913365  HAYASA COMERCIO E SERVICOS DE AUTOMOTORES LTDA   
3                11756671          YUKI NITEROI VEICULOS E SERVICOS LTDA.   
4                01976501                              A RIBEIRO FONTOURA   
...                   ...                                             ...   
387362           72320971          MARCHIORI MATERIAIS DE CONSTRUCAO LTDA   
387363           14108160                      LOTERICA ACERTO FACIL LTDA   
387364           08941799                     ELIANE FATIMA GOBETTI ZOTTI   
387365           87546479      POMPEIA COMERCIO DE VIDROS E MOLDURAS LTDA   
387366           19395513            ALAN CHRISTIAN BEATO CASSIANO EIRELI   

          Tipo   Ordem  MunicipioIBGE        Municipio_x  UF  \
0         Sede  I00001      3303302.0            NITEROI  RJ   
1       Filial  I00007      3303302.0            NITEROI  RJ   
2         Sede  I00001      3303302.0            NITEROI  RJ   
3         Sede  I00001      3303302.0            NITEROI  RJ   
4         Sede  I00001      3303302.0            NITEROI  RJ   
...        ...     ...            ...                ...  ..   
387362    Sede  I00001      4322186.0     TUPANCI DO SUL  RS   
387363    Sede  I00001      4322186.0     TUPANCI DO SUL  RS   
387364    Sede  I00001      4322186.0     TUPANCI DO SUL  RS   
387365    Sede  I00001      4314548.0     PINTO BANDEIRA  RS   
387366    Sede  I00001      5006275.0  PARAISO DAS AGUAS  MS   

                                  ServicosCorrespondentes     Posicao  \
0                                                  Inc. V  31/10/2019   
1                                                  Inc. V  31/10/2019   
2                                                  Inc. V  31/10/2019   
3                                                  Inc. V  31/10/2019   
4                                                  Inc. V  31/10/2019   
...                                                   ...         ...   
387362                  Inc. I, Inc. II, Inc. III, Inc. V  31/10/2019   
387363  Inc. I, Inc. II, Inc. III, Inc. IV, Inc. V, In...  31/10/2019   
387364                                           Inc. III  31/10/2019   
387365                                  Inc. II, Inc. III  31/10/2019   
387366  Inc. I, Inc. II, Inc. III, Inc. IV, Inc. V, In...  31/10/2019   

                                                  urlserv        Municipio_y  
0       https://was-p.bcnet.bcb.gov.br/olinda/servico/...            Niterói  
1       https://was-p.bcnet.bcb.gov.br/olinda/servico/...            Niterói  
2       https://was-p.bcnet.bcb.gov.br/olinda/servico/...            Niterói  
3       https://was-p.bcnet.bcb.gov.br/olinda/servico/...            Niterói  
4       https://was-p.bcnet.bcb.gov.br/olinda/servico/...            Niterói  
...                                  

In [14]:
dfnew.isna

<bound method DataFrame.isna of        CnpjContratante                                    NomeContratante  \
0             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
1             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
2             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
3             45441789    ADMINISTRADORA DE CONSORCIO NACIONAL HONDA LTDA   
4             36321990  AGORACRED S/A SOCIEDADE DE CRÉDITO, FINANCIAM...   
...                ...                                                ...   
387842        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387843        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387844        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387845        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   
387846        05192316  VIA CERTA FINANCIADORA S.A. - CRÉDITO, FINANC...   

       CnpjCorrespondente                  

In [12]:
dfnew.nunique()

CnpjContratante               278
NomeContratante               278
CnpjCorrespondente         145184
NomeCorrespondente         143696
Tipo                            4
Ordem                        9491
MunicipioIBGE                5570
Municipio                    5306
UF                             27
ServicosCorrespondentes        48
Posicao                         1
urlserv                         1
dtype: int64

In [ ]:
dfnew["Tipo"].value_counts()

In [ ]:
dfnew["Tipo"].value_counts().plot.bar()

In [ ]:
dfnew["NomeContratante"].value_counts().plot.barh(title = 'Nome Contratante')

In [ ]:
dfnew["UF"].value_counts(normalize=True)

In [ ]:
import folium
from folium.plugins import MarkerCluster

mapa = folium.Map(location=[45.523, -122.675],width=750, height=500)
#folium.Marker([-19.9166813,-43.9344931]).add_to(mapa)
mapa

In [ ]:
dfnew["UF"].value_counts().plot.map()

In [ ]:
dfnew["Municipio"].value_counts()

In [ ]:
dfnew.sort_values(by=["Tipo","ServicosCorrespondentes","Ordem","UF"])
dfnew()